design-proposals/vhostuser-interface.md

```diff
@@ -0,0 +1,96 @@
+- With VhostUser interface, the kubevirt VM created can utilize the dpdk virtio functionality and avoid using kernel path for the traffic when DPDK is involved, allowing usage of fast datapath and improve performance.
+
+## Motivation
+- Any kubevirt VM uses linux kernel interfaces to connect to its host and to reach out externally or to communicate within the cluster. The current kubevirt VMs can only support the kernel data path for which the traffic needs to be through the kernel of the node/compute hosting the VM.
+
+- As a result, when we have a host with DPDK dataplane running in Userspace, the traffic from kubevirt VM will go through the kernel and reach the dataplane running in userspace. This longer path slows down the traffic between the application and the dataplane when both are running in userspace.
+
+- In order to improve the traffic performance and make use of the dpdk, we need to use a different path that can by-pass the kernel and reach the dataplane from the user space directly. This so called fast path will run within the user space and will allow fast packet processing.
+
+## Goals
+- The kubevirt VM when created with DPDK features, should be able to use fast datpath if the host supports the DPDK dataplane allowing fast packet processing to improve performance of the traffic.
+
+## Non Goals
+- The user space networking will require the host to a DPDK dataplane.
+- The kubevirtci creates kernel mode dataplane hosts for testing the VMs. It needs to be updated to support DPDK.
+
+## Definition of Users
+- Any user who has a (k8s) cluster with a dpdk dataplane can use the interface/feature.
+- The feature will be supported only if a valid CNI is being used.
+
+## User Stories
+- As a user/admin, I want to use user space networking feature if I have DPDK supported VM and the host has a DPDK dataplane.
+
+## Repos
+
+- kubevirt/kubevirt
+   - Introduction of vhostuser type interface.
+- kubevirt/kubevirtci
+   - Enable usage of dpdk on the setups created by kubevirtci.
+
+# Design
+- The design involves introducing vHostUser interface type and required parameters to support it. An EmptyDIR volume (shared-dir) and DownwardAPI (podinfo) volume are mounted on the virtlauncher pod to have the support for new interface.
+- These mounts will allow the virt-launcher pod to create a VM with an additional interface which can be reached using the vhostSocket mentioned in the DownwardAPI.
+
+- Creating the VMs will follow the same process with a few changes highlighted as below:
+    1. **Once the VM spec virt-controller will add two new volumes to the virt-launcher pod.**\
+        a. **EmptyDir volume named (shared-dir) is used to share the vhostuser socket file with the virt-launcher pod and dpdk dataplane.**\
+        b. **DownwardAPI volume named (podinfo) is used to share vhostuser socket file name with the virt-launcher via pod annotations.**
```

Comment:
Is this required? What information will be provided this way?

The problem with `DownwardAPI` is that it is async and we can't really know when the information will be available in the context of `virt-launcher`. Therefore I would rather suggest that the `virt-controller` will proxy this information to `vmi` if we really need it. Then we can send a `sync` command at the right time.

Reply:
The DownwardAPI will provide the annotations containing the information on the socket file name and its location, which can be used by the virtualization-launcher pod while creating the VM using libvirt.

Reply:
Do you want to ever support migration? If so the (even without it) it would be required/preferable to not use the `DownwardAPI` rather us `virt-controller` which can propagate the annotations from `virt-launcher`  to `vmi`.

Reply:
> Do you want to ever support migration? If so the (even without it) it would be required/preferable to not use the `DownwardAPI` rather us `virt-controller` which can propagate the annotations from `virt-launcher` to `vmi`.

It is already in use for SR-IOV, by itself it is not a blocker for migration.
But if there are other reasonable options, I agree it would be preferable.

I mainly want to understand why it is needed in more details ([ref](https://github.com/kubevirt/community/pull/218#discussion_r1189408556)).